mm2src/trading_api/src/one_inch_api/client.rs

```diff
@@ -0,0 +1,164 @@
+            chain_id,
+            method_name,
+            query_params: vec![],
+        }
+    }
+
+    pub(crate) fn with_query_params(mut self, mut more_params: QueryParams<'a>) -> Self {
+        self.query_params.append(&mut more_params);
+        Self {
+            base_url: self.base_url,
+            endpoint: self.endpoint,
+            chain_id: self.chain_id,
+            method_name: self.method_name,
+            query_params: self.query_params,
+        }
+    }
+
+    #[allow(clippy::result_large_err)]
+    pub(crate) fn build(self) -> MmResult<Url, ApiClientError> {
+        let url = self
+            .base_url
+            .join(self.endpoint)?
+            .join(&format!("{}/", self.chain_id.to_string()))?
+            .join(self.method_name.as_str())?;
+        Ok(Url::parse_with_params(
+            url.as_str(),
+            self.query_params
+                .iter()
+                .map(|v| (v.0, v.1.as_str()))
+                .collect::<Vec<_>>(),
+        )?)
+    }
+}
+
+/// 1-inch API caller
+pub struct ApiClient {
+    base_url: Url,
+}
+
+impl ApiClient {
+    #[allow(unused_variables)]
+    #[allow(clippy::result_large_err)]
+    pub fn new(ctx: MmArc) -> MmResult<Self, ApiClientError> {
+        #[cfg(not(test))]
+        let url_cfg = ctx.conf["1inch_api"]
+            .as_str()
+            .ok_or(ApiClientError::InvalidParam("No API config param".to_owned()))?;
+
+        #[cfg(test)]
+        let url_cfg = ONE_INCH_API_TEST_URL;
+
+        Ok(Self {
+            base_url: Url::parse(url_cfg)?,
+        })
+    }
+
+    pub const fn eth_special_contract() -> &'static str { ONE_INCH_ETH_SPECIAL_CONTRACT }
+
+    pub const fn classic_swap_contract() -> &'static str { ONE_INCH_AGGREGATION_ROUTER_CONTRACT_V6_0 }
+
+    pub fn is_chain_supported(chain_id: u64) -> bool {
+        ONE_INCH_V6_0_SUPPORTED_CHAINS.iter().any(|(_name, id)| *id == chain_id)
+    }
+
+    fn get_headers() -> Vec<(&'static str, &'static str)> {
+        vec![
+            ("Authorization", ONE_INCH_API_TEST_AUTH.as_str()),
+            ("accept", "application/json"),
+        ]
+    }
+
+    fn get_swap_endpoint() -> &'static str { ONE_INCH_API_ENDPOINT_V6_0 }
+
+    pub const fn get_swap_method() -> &'static str { SWAP_METHOD }
+
+    pub const fn get_quote_method() -> &'static str { QUOTE_METHOD }
+
+    pub(crate) async fn call_api(api_url: Url) -> MmResult<Value, ApiClientError> {
+        let (status_code, _, body) = slurp_url_with_headers(api_url.as_str(), Self::get_headers())
+            .await
+            .mm_err(ApiClientError::HttpClientError)?;
+        let body = serde_json::from_slice(&body).map_to_mm(|err| ApiClientError::ParseBodyError(err.to_string()))?;
+        if status_code != StatusCode::OK {
+            let error = NativeError::new(status_code, body);
+            return Err(MmError::new(ApiClientError::from_native_error(error)));
+        }
+        Ok(body)
+    }
```

Comment:
we should delete json parsing to this function so it should be something like this

```suggestion
    pub(crate) async fn call_api<T:  Deserialize>(api_url: Url) -> MmResult<T, ApiClientError> {
        let (status_code, _, body) = slurp_url_with_headers(api_url.as_str(), Self::get_headers())
            .await
            .mm_err(ApiClientError::HttpClientError)?;
        let body = serde_json::from_slice(&body).map_to_mm(|err| ApiClientError::ParseBodyError(err.to_string()))?;
        if status_code != StatusCode::OK {
            let error = NativeError::new(status_code, body);
            return Err(MmError::new(ApiClientError::from_native_error(error)));
        }
        
        let data = serde_json::from_value::<T>(value).map_err(|err| ApiClientError::ParseBodyError(err.to_string()).into())
        Ok(data)
    }
```

Reply:
fixed in [23d8f1d](https://github.com/KomodoPlatform/komodo-defi-framework/pull/2222/commits/23d8f1d6e387b45d6f91ad37f411bb08e976c368), thanks